pkg/roachprod/vm/local/local.go

```diff
@@ -267,11 +276,33 @@ func (p *Provider) Create(
 }
 
 func (p *Provider) Grow(l *logger.Logger, vms vm.List, clusterName string, names []string) error {
-	return errors.New("unimplemented")
+	now := timeutil.Now()
+	offset := p.clusters[clusterName].VMs.Len()
+	for i := range names {
+		cVM, err := p.createVM(clusterName, i+offset, now)
+		if err != nil {
+			return err
+		}
+		p.clusters[clusterName].VMs = append(p.clusters[clusterName].VMs, cVM)
+	}
+	return p.storage.SaveCluster(l, p.clusters[clusterName])
 }
 
 func (p *Provider) Shrink(l *logger.Logger, vmsToDelete vm.List, clusterName string) error {
-	return errors.New("unimplemented")
+	keepCount := p.clusters[clusterName].VMs.Len() - len(vmsToDelete)
+	keepVMs := make(vm.List, 0, keepCount)
+	for i := 0; i < p.clusters[clusterName].VMs.Len(); i++ {
+		if i < keepCount {
+			keepVMs = append(keepVMs, p.clusters[clusterName].VMs[i])
```

Comment:
Is there a reason we can't create a new slice, backed by the same underlying array? I.e., why do we need a new copy?

```go
p.clusters[clusterName].VMs = p.clusters[clusterName].VMs[:keepCount]
```

And then remove the dirs for `p.clusters[clusterName].VMs[keepCount:]`. Let me know if I'm missing something.

Reply:
True. no need to build a new slice. Updated.